File: 09_DataStructure_Algorithm/0519_day52/2_전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    answer = n
    
    def dfs(cur):
        visited[cur] = True
        count = 1
        
        for nxt in graph[cur]:
            if not visited[nxt]:
                count += dfs(nxt)
        
        return count
        
    for i in range(n - 1):
        graph = [[] for _ in range(n + 1)]
        visited = [False] * (n + 1)
        for j in range(n - 1):
            if j != i:
                graph[wires[j][0]].append(wires[j][1])
                graph[wires[j][1]].append(wires[j][0])
        
        answer = min(answer, abs(dfs(wires[i][0]) - dfs(wires[i][1])))
    
    return answer

##
#
# 해설집 참고

def solution(n, wires):
    graph = [[] for _ in range(n + 1)]
    
    for w1, w2 in wires:
        graph[w1].append(w2)
        graph[w2].append(w1)
        
    visited = [False] * (n + 1)
    subtrees = [[] for _ in range(n + 1)]
    
    def dfs(cur):
        visited[cur] = True
        subtrees[cur] = 1
        
        for nxt in graph[cur]:
            if not visited[nxt]:
                dfs(nxt)
                subtrees[cur] += subtrees[nxt]
                
    dfs(1)
    
    answer = float("INF")
    for i in range(2, n + 1):
        answer = min(answer, abs(n - 2 * subtrees[i]))
    
    return answer

File: 09_DataStructure_Algorithm/0519_day52/2_전력망을 둘로 나누기/전력망을 둘로 나누기.java
// 완전 탐색으로 풀기
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int answer = n;
        
        for (int k = 0; k < n - 1; k++) {
            List<List<Integer>> tree = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                tree.add(new ArrayList<>()) ;
            }
            for (int j = 0; j < n - 1; j++) {
                if (k == j) continue;
                tree.get(wires[j][0] - 1).add(wires[j][1] - 1);
                tree.get(wires[j][1] - 1).add(wires[j][0] - 1);
            }
            
            int count = dfs(0, tree, new boolean[n]);
            answer = Math.min(answer, Math.abs(2 * count - n));
        }
        
        return answer;
    }
    
    int dfs(int cur, List<List<Integer>> tree, boolean[] visited) {
        int count = 1;
        visited[cur] = true;
        
        for (int child : tree.get(cur)) {
            if (!visited[child]) {
                count += dfs(child, tree, visited);
            }
        }
    
        return count;
    }
}
// => O(n²)

// DFS
import java.util.*;

class Solution {
    int minDiff;
    public int solution(int n, int[][] wires) {
        minDiff = n;
        
        // 간선리스트 => 인접리스트 변환
        List<List<Integer>> tree = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            tree.add(new ArrayList<>()) ;
        }
        for (int[] e : wires) {
            tree.get(e[0] - 1).add(e[1] - 1);
            tree.get(e[1] - 1).add(e[0] - 1);
        }
        
        dfs(0, tree, new boolean[n]);
        
        return minDiff;
    }
    
    // 노드의 개수를 세면서 차이값 계산해서 업데이트
    int dfs(int root, List<List<Integer>> tree, boolean[] visited) {
        visited[root] = true;
        int count = 1;
        
        for (int child : tree.get(root)) {
            if (!visited[child]) {
                count += dfs(child, tree, visited);
            }
        }
        
        int cnt1 = count;
        int cnt2 = tree.size() - count;
        
        minDiff = Math.min(minDiff, Math.abs(cnt1 - cnt2));
        
        return count;
    }
}